tsdb/wal/wal_test.go

```diff
@@ -204,6 +204,56 @@ func TestWALRepair_ReadingError(t *testing.T) {
 	}
 }
 
+// TestUnsequentialSegments tests whether the Unsequential wal segments are
+// deleted or not.
+func TestUnsequentialSegments(t *testing.T) {
+	dir, err := ioutil.TempDir("", "wal_unsequential")
+	testutil.Ok(t, err)
+	defer func() {
+		testutil.Ok(t, os.RemoveAll(dir))
+	}()
+
+	var (
+		logger         = testutil.NewLogger(t)
+		segmentSize    = pageSize * 3
+		recordSize     = pageSize - recordHeaderSize
+		segments       = 20
+		deletePosition = segments / 2
+	)
+
+	// Create temporary segments
+	{
+		for i := 0; i < segments; i++ {
+			w, err := NewSize(logger, nil, dir, segmentSize, false)
+			testutil.Ok(t, err)
+			buf := make([]byte, recordSize)
+			_, err = rand.Read(buf)
+			testutil.Ok(t, err)
+			testutil.Ok(t, w.Log(buf))
+		}
+	}
+	f, err := ioutil.ReadDir(dir)
+	testutil.Ok(t, err)
+
+	// Check if the total created segments have number equal to the segments
+	testutil.Assert(t, len(f) == segments, "Segments not created as expected")
+
+	// Delete a segment in order to create a gap or unsequential
+	deletionSegName := f[deletePosition].Name()
+	correctSegmentsCount := segments - deletePosition
+
+	// Create a gap in the middle of the sequence
+	os.Remove(dir + "/" + deletionSegName)
+	w, err := Open(logger, nil, dir)
+	testutil.Ok(t, err)
+
+	w.RepairUnsequentialSegments(dir)
+
+	df, err := ioutil.ReadDir(dir)
+	testutil.Ok(t, err)
+	testutil.Assert(t, len(df) == correctSegmentsCount, "Unable to repair unsequential segments")
```

Comment:
We also need to verify that the sequence is fine after the repair and that `NewSize` returns no error.

Additionally, verify which segments are present after the repair according to the expected behaviour. The number of segments would not be enough.